### PR TITLE
plat: Migrate bus drivers to drivers/ukbus

### PR DESCRIPTION
**Pull request by michpappas**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Description of changes

Migrate the PCI and platform drivers to `drivers/ukbus/`

GitHub-Depends-On: #1023

### CODE REVIEW

**Comment by michpappas** [comment]
~~Notice that the driver is introduced as a platform library, ie:~~
```
$(eval $(call addplatlib_s,kvm,libukbus_pci,$(CONFIG_LIBUKBUS_PCI)))
```
~~This is due to limitations of #1023. **Do not merge until resolved.**~~

**Comment by michpappas** [comment]
Changes in this update:
- Rebase to latest state of #1023
- Replace subsystem's `Config.uk` `menuconfig` with `menu`
- Register library using `addlib_s`
- Add `exportsyms.uk`
- Add `ASINCLUDES` and `CSSINCLUDES`

File: drivers/ukbus/pci/arch/arm64/pci_bus.c
Comment (mogasergiu) [inline drivers/ukbus/pci/arch/arm64/pci_bus.c]: Perhaps an `unlikely` here? :thinking:

File: drivers/ukbus/pci/pci_ecam.c
Comment (mogasergiu) [inline drivers/ukbus/pci/pci_ecam.c:237]: commit message headline too long, it appears as
```
drivers/ukbus/pci: Fix initialization of dummy_imask in gen_pci_irq_p…
…arse
```

I see that you added
```
Checkpatch-Ignore: COMMIT_LOG_LONG_LINE
```
Maybe it would be worth re-wording it though as this is too generic and not specific enough. How about:
```
drivers/ukbus/pci: Memset `dummy_imask` by size not unit count
```
I've also been met with similar situations quite frequently unfortunately :disappointed:.

File: plat/kvm/Config.uk
Comment (mogasergiu) [inline plat/kvm/Config.uk:15]: Why `if ARCH_X86_64`?

**Review by mogasergiu — COMMENTED** [review]
Had a quick run, will do a round of testing as well after my questions have been addressed and another for reviewing as well. Otherwise I greatly appreciate the fact that you did take my opinion into account regarding commit layout, thank you :pray:.

File: drivers/ukbus/pci/pci_ecam.c
Comment (michpappas) [reply drivers/ukbus/pci/pci_ecam.c:237]: Actually the headline is 73 chars, while the checkpatch limit is 75. The problem is the error message of GCC. I will wrap the long line (~120 chars) so that the max line is 80 chars.

File: plat/kvm/Config.uk
Comment (michpappas) [reply plat/kvm/Config.uk:15]: This is part of a different PR, I only fixed indentation here. But since you ask: Normally the platform would choose a specific driver (eg PIC, x2APIC, GICv2, GICv3), which in turn would ~~select~~ `depend on` the API library, ie LIBUKINTCTLR. However, right now on x86 we don't support this yet, so we enable the library that provides the API. And on Arm we need to select the GIC driver at the VMM level, as QEMU virt by default uses GICv2 while FC uses GICv3.

File: drivers/ukbus/pci/pci_ecam.c
Comment (mogasergiu) [reply drivers/ukbus/pci/pci_ecam.c:237]: > Actually the headline is 73 chars, while the checkpatch limit is 75.

Which is still above 70, the maximum recommended for headlines AFAIK. Also, it is too generic, but I do not have any strong feelings towards this, keep it as you like.

> The problem is the error message of GCC. I will wrap the long line (~120 chars) so that the max line is 80 chars.

I would actually leave that as is. From what I have seen on mailing lists, usually errors/warnings/crashdumps/etc are kept without wrapping

File: plat/kvm/Config.uk
Comment (mogasergiu) [reply plat/kvm/Config.uk:15]: Ah, I see makes perfect sense, so this is temporary for x86 :+1:

File: drivers/ukbus/pci/pci_ecam.c
Comment (michpappas) [reply drivers/ukbus/pci/pci_ecam.c:237]: This is the relevant snippet from checkpatch:
```
# Check for line lengths > 75 in commit log, warn once
                if ($in_commit_log && !$commit_log_long_line &&
                    length($line) > 75 &&
                    !($line =~ /^\s*[a-zA-Z0-9_\/\.]+\s+\|\s+\d+/ ||
                                        # file delta changes
                      $line =~ /^\s*(?:[\w\.\-]+\/)++[\w\.\-]+:/ ||
                                        # filename then :
                      $line =~ /^\s*(?:Fixes:|Link:)/i ||
                                        # A Fixes: or Link: line
                      $commit_log_possible_stack_dump)) {
                        WARN("COMMIT_LOG_LONG_LINE",
                             "Possible unwrapped commit description (prefer a maximum 75 chars per line)\n" . $herecurr);
                        $commit_log_long_line = 1;
                }
```

>  I would actually leave that as is. From what I have seen on mailing lists, usually errors/warnings/crashdumps/etc are kept without wrapping

Yes, that was my initial rationale. Ok then I'll leave it and we can blame GCC for spewing such long messages.

File: drivers/ukbus/pci/Makefile.uk
Comment (rares-miculescu) [inline drivers/ukbus/pci/Makefile.uk:7]: I think you should also include the platform headers, because `pci_ecam.c` has `#include <uk/bus/platform.h>`.

**Review by rares-miculescu — APPROVED** [review]
All good on my side. Thank you, @michpappas!

Reviewed-by: Rares Miculescu [miculescur@gmail.com](mailto:miculescur@gmail.com)

**Comment by michpappas** [comment]
Pushed an update to address an issue identified by @rares-miculescu where enabling `CONFIG_LIBFDT` on x86 causes a compile error:

```
/home/mpp/devel/unikraft/uk_helloworld/unikraft/drivers/ukbus/pci/pci_ecam.c:34:10: fatal error: uk/bus/platform.h: No such file or directory
   34 | #include <uk/bus/platform.h>
      |          ^~~~~~~~~~~~~~~~~~~
```

The issue is caused by:
1. `libfdt` being visible on `x86_64`
2. `pci_ecam.c` incorrectly depending only on `CONFIG_LIBFDT`, given that it also depends on `CONFIG_LIBUKBUS_PLATFORM` (which is only enabled on Arm by the platform).

I pushed an update that introduces an invisible option `LIBUKBUS_PCI_ECAM` that is enabled conditionally to the above dependencies.

File: drivers/ukbus/pci/pci_ecam.c
Comment (mogasergiu) [inline drivers/ukbus/pci/pci_ecam.c:237]: For the commit message:
`s/initializaztion/initialization/`

File: drivers/ukbus/pci/arch/arm64/pci_bus.c
Comment (mogasergiu) [inline drivers/ukbus/pci/arch/arm64/pci_bus.c]: I know this was present here before, but the only other `checkpatch.pl --strict` `CHECK` I am getting in this PR is regarding the whitespace after the cast. I know this is going to be deleted anyway when it will be reworked so probably no point in pointing this out now, but I thought I'd let you know in case that maybe you missed this and you would have wanted to modify it regardless. I leave it up to you if you want to waste some more time modifying this or not.

**Review by mogasergiu — APPROVED** [review]
This all looks good to me. Just two tiny comments. Nevertheless, this is great important work that I can't wait to get merged in its entirety. Make Unikraft great again, driver infrastructure coming in! :truck: 

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**
